### PR TITLE
Build stuff (wip) review

### PR DESCRIPTION
**Pull request by marcdumais-work**
Salut Antoine,

Can you have a look at what I have for build so far? It's all in the backend : 
- a build output parser, that parses for GCC errors/warnings
- a build launcher, that starts a process with a given name/cwd. The build parser is hooked-up in there, and parses the build output as it happens, emitting any error/warning it finds. 

Both have unit test that can help understand how they work. 

P.S. already I think we could improve the build output parser. Now I use the  "gcc-output-parser" npm package, but it would probably be better to make our own, re-using the regexp from CDT. It would simplify it a bit too, since the CDT parser is going at it line-by-line instead of having multi-line expressions.

### CODE REVIEW

**Comment by svenefftinge** [comment]
Did you see my question regarding the build component (#183) vs. the launch configs (#184)?
Is that the same, or do you want to have a separated component for builds?
https://github.com/theia-ide/theia/issues/183#issuecomment-311570933

File: src/builds/node/gcc-build-parser.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    getDefaultChunkOverlapValue(): number {
+        return this.myChunkOverlap
+    }
+
+    doParse(buffer: Buffer, indexOffset: number): IBuildOutputParsedEntry[] {
+        let parsedEntries: IBuildOutputParsedEntry[] = []
+
+        // parse gcc output using "gcc-output-parser"
+        let entries: IGccBuildOutputParsedEntry[] = gccOutputParser.parseString(buffer)
+
```
Comment (hexa00) [inline src/builds/node/gcc-build-parser.ts]: Seems like maybe the chunkoverlap is a limitaiton of the gccoutputparser... I think you wanted to get rid of that anyway ?

File: src/builds/node/launch-build.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import { GccBuildParser } from './gcc-build-parser';
+import * as Events from "events"
+import * as child from 'child_process';
+
+
+export interface LaunchBuildinfo {
+    /** the directory where to execute the build */
+    cwd: string
+    /** build executable or script to call */
+    buildCommand: string,
```
Comment (hexa00) [inline src/builds/node/launch-build.ts]: Command infers to me that I may be able to send arguments directly there.. should it be more like buildExecutablePath or something ?

File: src/builds/node/launch-build.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+export class LaunchBuild extends Events.EventEmitter implements ILaunchBuild {
+
+    public launch(launchinfo: LaunchBuildinfo): Promise<string> {
+        let build: child.ChildProcess
+
+        return new Promise((resolve, reject) => {
+            build = child.spawn(launchinfo.buildCommand, launchinfo.buildCommandArgs || [], { cwd: launchinfo.cwd })
+
+            let parser: GccBuildParser = new GccBuildParser(build.stderr)
+
```
Comment (hexa00) [inline src/builds/node/launch-build.ts]: This should be a dependency of LaunchBuild with inversify

File: src/builds/node/launch-build.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+export interface ILaunchBuild {
+    launch(launchinfo: LaunchBuildinfo): Promise<string>
+}
+
+export interface ILaunchResult {
+    stdout: NodeJS.ReadableStream,
+    stderr: NodeJS.ReadableStream
+}
+
+export class LaunchBuild extends Events.EventEmitter implements ILaunchBuild {
```
Comment (hexa00) [inline src/builds/node/launch-build.ts]: LaunchBuild sounds like this should be 2 different things Launch and Build and like Launch should get Build as argument or something or Build just be something more generic as a dependency. Maybe Launch should have LaunchContributions and Build should be one of them.. then you could have like Launch -> execute Build -> execute -> Run etc... (Something to discuss)

I'm also working on making something that exposes a terminal more generic so we should find some synergies there.

File: src/builds/node/launch-build.ts (hunk 2 above)
Comment (hexa00) [inline src/builds/node/launch-build.ts]: I just related that this is the naming using in nodejs itself so I guess it's OK

File: npm-shrinkwrap.json [hunk 5]
```diff
@@ -123,7 +123,7 @@
     },
     "@types/express-serve-static-core": {
       "version": "4.0.46",
-      "from": "@types/express-serve-static-core@4.0.46",
+      "from": "@types/express-serve-static-core@*",
       "resolved": "https://registry.npmjs.org/@types/express-serve-static-core/-/express-serve-static-core-4.0.46.tgz"
```
Comment (marcdumais-work) [inline npm-shrinkwrap.json]: hummm. Strange indeed. It disapeared when I rebased my patches.

Comment (marcdumais-work) [inline npm-shrinkwrap.json]: no, still there...

Comment (marcdumais-work) [inline npm-shrinkwrap.json]: fixed

File: src/builds/node/abstract-build-parser.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+     * largest expected size for an entry.
+     */
+    abstract getDefaultChunkOverlapValue(): number
+
+    /**
+     * Parse for errors, warnings, etc in the buffer given as parameter. Returns
+     * an array of IBuildOutputParsedEntry, one entry per error, warning, note found.
+     *
+     * @param buffer - Buffer containing potential error, warning, ... entries
+     * @param: indexOffset - offset of the current chunk from the beginning of the log
```
Comment (marcdumais-work) [inline src/builds/node/abstract-build-parser.ts]: done

File: src/builds/node/abstract-build-parser.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+    public parse(): Promise<IParsedBuildLog> {
+
+        this.chunkOverlap = this.getDefaultChunkOverlapValue()
+
+        let offset = 0
+        return new Promise<IParsedBuildLog>((resolve, reject) => {
+            try {
+                let firstIteration: boolean = true
+                let stream: NodeJS.ReadableStream = this.inputStream
```
Comment (marcdumais-work) [inline src/builds/node/abstract-build-parser.ts]: done

File: src/builds/node/build-parser.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+export interface IBuildLogParser {
+    /** starts the parsing */
+    parse(): Promise<IParsedBuildLog>
```
Comment (marcdumais-work) [inline src/builds/node/build-parser.ts]: These are JSDoc documentation entries, similar to JavaDoc. It permits having those comments be displayed when hovering over the fields.

File: src/builds/node/gcc-build-parser.spec.ts [hunk 9]
```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { GccBuildParser } from "./gcc-build-parser";
+import * as Stream from "stream";
```
Comment (marcdumais-work) [inline src/builds/node/gcc-build-parser.spec.ts]: done

File: src/builds/node/build-parser.ts (hunk 8 above)
Comment (hexa00) [inline src/builds/node/build-parser.ts]: ha ok good, I'll see if I can use that too

File: src/builds/node/abstract-build-parser.ts (hunk 6 above)
Comment (marcdumais-work) [inline src/builds/node/abstract-build-parser.ts]: done

File: src/builds/node/abstract-build-parser.ts [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+                    }
+
+                    for (let parsedEntry of parsedEntries) {
+                        // avoid duplicated entries
+                        if (!this.isEntryKnown(parsedEntry)) {
+                            this.parsedBuildLog.entries.push(parsedEntry)
+                            try {
+                                this.emit('new-entry', parsedEntry)
+                            } catch (err) {
+                                console.log("Error caught while emitting: " + err)
```
Comment (marcdumais-work) [inline src/builds/node/abstract-build-parser.ts]: done

**Comment by marcdumais-work** [comment]
this PR is obsolete. see instead https://github.com/theia-ide/theia/pull/389

**Pull request closed without merging**